File: Java/Tree/Tree_06.java
/* 
=============================================================================== 
Program Name    : Binary Search Tree - Search Element 
Program Number  : 06 
File Name       : Tree_06.java 
 
Description     : This program demonstrates how to search an element in a 
                  Binary Search Tree using Java. It creates a Binary Search 
                  Tree by inserting elements and searches for a given element 
                  using the Binary Search Tree property. 
 
Date            : August 2026 
Language        : Java 
Author          : Ishwari Vijaykumar Surve 
=============================================================================== 
 
Concepts Used: 
1. Class 
2. Object-Oriented Programming (OOP) 
3. Self-Referential Class 
4. Constructor 
5. Binary Search Tree (BST) 
6. Dynamic Memory Allocation 
7. Insertion Operation 
8. Searching Operation 
9. Boolean Return Value 
10. Iterative Searching 
 
Key Learning Points: 
- Understand how searching is performed in a Binary Search Tree. 
- Learn how the Binary Search Tree property makes searching efficient. 
- If the element is greater than the current node, move to the right subtree. 
- If the element is smaller than the current node, move to the left subtree. 
- If the element is equal to the current node, the element is found. 
- Understand the use of boolean return values. 
 
=============================================================================== 
*/ 
 
class node 
{ 
    public int data; 
    public node lchild; 
    public node rchild; 
} 
 
class Tree 
{ 
    private node first; 
 
    public Tree() 
    { 
        first = null; 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Insert 
    // Description   : Insert element into Binary Search Tree 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public void Insert(int iNo) 
    { 
        node newn = null; 
        node temp = null; 
 
        newn = new node(); 
 
        newn.data = iNo; 
        newn.lchild = null; 
        newn.rchild = null; 
 
        if(first == null) 
        { 
            first = newn; 
        } 
        else 
        { 
            temp = first; 
 
            while(true) 
            { 
                if(iNo > temp.data) 
                { 
                    if(temp.rchild == null) 
                    { 
                        temp.rchild = newn; 
                        break; 
                    } 
 
                    temp = temp.rchild; 
                } 
                else if(iNo < temp.data) 
                { 
                    if(temp.lchild == null) 
                    { 
                        temp.lchild = newn; 
                        break; 
                    } 
 
                    temp = temp.lchild; 
                } 
                else 
                { 
                    System.out.println("Duplicate element not allowed"); 
                    break; 
                } 
            } 
        } 
    } 
 
    /////////////////////////////////////////////////////////////////////////// 
    // 
    // Function Name : Search 
    // Description   : Search element in Binary Search Tree 
    // 
    /////////////////////////////////////////////////////////////////////////// 
 
    public boolean Search(int iNo) 
    { 
        node temp = null; 
 
        temp = first; 
 
        while(temp != null) 
        { 
            if(iNo == temp.data) 
            { 
                return true; 
            } 
            else if(iNo > temp.data) 
            { 
                temp = temp.rchild; 
            } 
            else 
            { 
                temp = temp.lchild; 
            } 
        } 
 
        return false; 
    } 
} 
 
class Tree_06 
{ 
    public static void main(String A[]) 
    { 
        Tree tobj = new Tree(); 
 
        tobj.Insert(11); 
        tobj.Insert(5); 
        tobj.Insert(17); 
        tobj.Insert(4); 
        tobj.Insert(7); 
        tobj.Insert(15); 
        tobj.Insert(21); 
 
        int iNo = 15; 
 
        System.out.println("Searching Element : " + iNo); 
 
        if(tobj.Search(iNo) == true) 
        { 
            System.out.println("Element is Present"); 
        } 
        else 
        { 
            System.out.println("There is no such element"); 
        } 
    } 
} 
 
/* 
=============================================================================== 
Expected Output: 
=============================================================================== 
 
Searching Element : 15 
Element is Present 
 
=============================================================================== 
Memory Representation: 
=============================================================================== 
 
                         11 
                       /    \ 
                      5      17 
                     / \    /  \ 
                    4   7  15  21 
 
Searching for : 15 
 
                         11 
                           \ 
                           17 
                          / 
                        15 
 
Search Path: 
 
15 > 11  -> Move Right 
15 < 17  -> Move Left 
15 = 15  -> Element Found 
 
=============================================================================== 
Search Process: 
=============================================================================== 
 
Step 1 : Start from the first node 11. 
Step 2 : Compare 15 with 11. 
Step 3 : Since 15 is greater than 11, move to the right child. 
Step 4 : Reach node 17. 
Step 5 : Compare 15 with 17. 
Step 6 : Since 15 is smaller than 17, move to the left child. 
Step 7 : Reach node 15. 
Step 8 : Since 15 is equal to 15, the element is found. 
 
=============================================================================== 
Node Structure: 
=============================================================================== 
 
                    +-----------+-----------+-----------+ 
                    |   data    |  lchild   |  rchild   | 
                    +-----------+-----------+-----------+ 
                          |           |           | 
                        Value       Address     Address 
 
=============================================================================== 
Algorithm: 
=============================================================================== 
 
1. Start the program. 
2. Define the node class with data, lchild and rchild members. 
3. Define the Tree class. 
4. Initialize first to NULL using the constructor. 
5. Define the Insert function. 
6. Insert the given elements into the Binary Search Tree. 
7. Define the Search function. 
8. Start searching from the first node. 
9. Compare the given element with the current node. 
10. If both elements are equal, return true. 
11. If the given element is greater, move to the right subtree. 
12. If the given element is smaller, move to the left subtree. 
13. Continue until the element is found or the current node becomes NULL. 
14. If the current node becomes NULL, return false. 
15. Display whether the element is present or not. 
16. Stop the program. 
 
=============================================================================== 
Program Flow: 
=============================================================================== 
 
        Start 
          | 
          v 
     Create Tree 
          | 
          v 
   Insert Tree Elements 
          | 
          v 
    Set iNo = 15 
          | 
          v 
     Call Search() 
          | 
          v 
   Start from first 
          | 
          v 
    Compare Element 
       /    |    \ 
   Equal  Greater  Smaller 
     |       |       | 
     v       v       v 
   Found    Right   Left 
             |       | 
             +---+---+ 
                 | 
                 v 
          Continue Search 
                 | 
                 v 
         Display Result 
                 | 
                 v 
                Stop 
 
=============================================================================== 
Time Complexity: 
=============================================================================== 
 
Best Case    : O(1) 
Average Case : O(log n) 
Worst Case   : O(n) 
 
=============================================================================== 
Space Complexity: 
=============================================================================== 
 
Tree Storage    : O(n) 
Search Variable : O(1) 
 
=============================================================================== 
Conclusion: 
=============================================================================== 
 
Program 06 demonstrates the searching operation in a Binary Search Tree 
using Java. The Search function compares the given element with each node 
and moves either to the left or right subtree according to the BST property. 
The function returns true when the element is found and false when the 
element is not present in the Tree. 
 
=============================================================================== 
*/

File: C++/Tree/Tree_06.cpp
/*
===============================================================================
Program Name    : Binary Search Tree - Postorder Traversal
Program Number  : 06
File Name       : Tree_06.cpp 
Description     : This program implements Postorder Traversal of a Binary
                  Search Tree using C++. Postorder Traversal visits the
                  nodes in Left, Right, Data (LRD) order.

Date            : August 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Binary Search Tree (BST)
2. Postorder Traversal
3. Recursion
4. Class and Object
5. Self-Referential Structure
6. Pointers

Key Learning Points:
- Understand Postorder Traversal of a Binary Search Tree.
- Learn the LRD traversal technique.
- Understand recursive traversal of Tree nodes.
- Learn how the root node is processed after its subtrees.
- Understand the use of a wrapper function for traversal.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *lchild;
    struct node *rchild;
};

class BST
{
private:
    struct node *first;

public:
    BST();
    void Insert(int iNo);

    void Postorder(struct node *first);
    void Display();
};

// Constructor
BST::BST()
{
    first = NULL;
}

// Insert Node
void BST::Insert(int iNo)
{
    struct node *newn = NULL;
    struct node *temp = NULL;

    newn = new node;

    newn->data = iNo;
    newn->lchild = NULL;
    newn->rchild = NULL;

    if(first == NULL)
    {
        first = newn;
    }
    else
    {
        temp = first;

        while(1)
        {
            if(iNo > temp->data)
            {
                if(temp->rchild == NULL)
                {
                    temp->rchild = newn;
                    break;
                }

                temp = temp->rchild;
            }
            else if(iNo < temp->data)
            {
                if(temp->lchild == NULL)
                {
                    temp->lchild = newn;
                    break;
                }

                temp = temp->lchild;
            }
            else
            {
                cout<<"Duplicate element not allowed\n";
                delete newn;
                break;
            }
        }
    }
}

// Postorder Traversal
// L R D
void BST::Postorder(struct node *first)
{
    if(first != NULL)
    {
        Postorder(first->lchild);

        Postorder(first->rchild);

        cout<<first->data<<"\n";
    }
}

// Wrapper Function
void BST::Display()
{
    Postorder(first);
}

int main()
{
    BST bobj;

    bobj.Insert(11);
    bobj.Insert(5);
    bobj.Insert(17);
    bobj.Insert(4);
    bobj.Insert(7);
    bobj.Insert(15);
    bobj.Insert(21);

    cout<<"Postorder Traversal :\n";

    bobj.Display();

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Postorder Traversal :
4
7
5
15
21
17
11

===============================================================================
Memory Representation:
===============================================================================

                         first
                           |
                           ▼
                       +-------+
                       |  11   |
                       +---+---+
                          / \
                         /   \
                        ▼     ▼
                    +-----+ +-----+
                    |  5  | | 17  |
                    +--+--+ +--+--+
                      / \     / \
                     /   \   /   \
                    ▼     ▼ ▼     ▼
                  +---+ +---+ +---+ +---+
                  | 4 | | 7 | |15 | |21 |
                  +---+ +---+ +---+ +---+

Postorder Traversal: Left → Right → Data

Traversal Sequence:

4 → 7 → 5 → 15 → 21 → 17 → 11

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create a Binary Search Tree.
3. Insert the required elements into the BST.
4. Call the Display() function.
5. Display() calls the Postorder() function with the root node.
6. In Postorder(), recursively visit the left subtree.
7. Recursively visit the right subtree.
8. Display the data of the current node.
9. Continue the process until all nodes are visited.
10. Stop the program.

===============================================================================
Program Flow:
===============================================================================

                         Start
                           │
                           ▼
                    Create BST Object
                           │
                           ▼
                    Insert BST Nodes
                           │
                           ▼
                     Call Display()
                           │
                           ▼
                   Call Postorder()
                           │
                           ▼
                  Traverse Left Subtree
                           │
                           ▼
                 Traverse Right Subtree
                           │
                           ▼
                    Display Node Data
                           │
                           ▼
                    All Nodes Visited?
                       /          \
                     No            Yes
                     │              │
                     └─── Repeat    ▼
                                  Stop

===============================================================================
Time Complexity:
===============================================================================

Insert() - Best Case    : O(1)
Insert() - Average Case : O(log n)
Insert() - Worst Case   : O(n)

Postorder() - Best Case   : O(n)
Postorder() - Average Case: O(n)
Postorder() - Worst Case  : O(n)

Where n is the number of nodes in the BST.

===============================================================================
Space Complexity:
===============================================================================

Postorder() Auxiliary Space : O(h)

Where h is the height of the Tree.

For a balanced Tree : O(log n)
For a skewed Tree   : O(n)

Total Space : O(n)

===============================================================================
Conclusion:
===============================================================================

Program 006 implements Postorder Traversal of a Binary Search Tree using
C++. The traversal follows the Left, Right, Data (LRD) order. The left
subtree and right subtree are processed before the current node. Recursion
is used to traverse the complete Tree.

===============================================================================
*/
